### ISSUE TITLE
Passing lambdas to algorithm interfaces (for_each, generate, min_element, ...)

### ISSUE THREAD

**Issue by devreal** [dash-project/dash#274]
I noticed that the signature of `dash::for_each` requires a `std::function<void(ElementType &)>` to be passed in as functor. In particular, it seems that this excludes passing lambdas instead of pre-defined function arguments. The following fails to compile with the error that the lambda is not derived from `std::function`:

```
dash::Array<int> arr(dash::size());
dash::for_each(arr.begin(), arr.end(), [](int& val){ val = dash::size(); })
```

Is there any reason for explicitly excluding lambdas here? On my system, the function type is part of the template specification:

```
  template<typename _InputIterator, typename _Function>
    _Function
    for_each(_InputIterator __first, _InputIterator __last, _Function __f) { ... }
```

**Comment by rkowalewski** [comment]
I agree with this limitation. Moreover, lambdas are more efficient in many cases compared to `std::function` as they can be better optimized by the compiler.

**Comment by fuchsto** [comment]
I agree, there was no specific reason against using a template type like `UnaryFunction`, I just didn't think of it back then.

**Comment by fuchsto** [comment]
Fixed in #275
